fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/rules/AbstractDataAccessRule.java

```diff
@@ -227,31 +232,187 @@ protected Set<? extends RelationalExpression> dataAccessForMatchPartition(@Nonnu
+
+            int numCombinations = 0;
+            int numDiscardedCombinations = 0;
+
+            // create intersections for all n choose k partitions from k = 3 ... n
+            for (int k = 3; k < bestMaximumCoverageMatches.size() + 1; k ++) {
+                boolean hasCommonOrderingForK = false;
+                for (final var kPartition : ChooseK.chooseK(bestMaximumCoverageMatches, k)) {
+                    numCombinations ++;
+                    call.emitEvent(Debugger.Location.ALL_INTERSECTION_COMBINATIONS);
+
+                    //
+                    // For a combination of n orderings we enumerate find all the positions in that the combination
+                    // does not contain. Starting from a matrix where all bits are set we clear all columns and
+                    // rows at those positions. That is the check bit matrix.
+                    //
+                    setAll(checkBitMatrix); // all bits to 1
+                    final var membership = new BitSet(bestMaximumCoverageMatches.size());
+                    for (final var singleMatchedAccessVectored : kPartition) {
+                        membership.set(singleMatchedAccessVectored.getPosition());
+                    }
+
+                    for (int position = membership.nextClearBit(0);
+                             position >= 0 && position < bestMaximumCoverageMatches.size();
+                             position = membership.nextClearBit(position + 1)) {
+                        clearRowAndColumnAtPosition(checkBitMatrix, position);
+                    }
+
+                    if (!hasCommonOrdering(sieveBitMatrix, checkBitMatrix)) {
+                        numDiscardedCombinations ++;
+                        call.emitEvent(Debugger.Location.DISCARDED_INTERSECTION_COMBINATIONS);
+                        continue;
+                    }
+
+                    final var intersectionResult =
                             createIntersectionAndCompensation(
                                     call,
                                     commonPrimaryKeyValues,
                                     bestMatchToDistinctPlanMap,
-                                    partition,
-                                    requestedOrderings).stream())
-                    .forEach(resultSet::add);
-        });
+                                    kPartition,
+                                    requestedOrderings);
+
+                    Verify.verify(intersectionResult.hasCommonIntersectionOrdering());
+                    hasCommonOrderingForK = true;
+
+                    resultSet.addAll(intersectionResult.getExpressions());
+                }
+
+                if (!hasCommonOrderingForK) {
+                    // if we weren't able to find combinations of size k that created intersections, there won't be any
+                    // of size k + 1
+                    break;
+                }
```

Comment:
This is a nice optimisation.
I wonder if we can integrate the "forbidden combinations" concept in the chooseK iterator logic itself, I would argue it would make it even more efficient (and perhaps elegant) considering the following interface:

```java
chooseK(N,K, sieveMatrix)
```

Then, there will be a pruning logic, similar to the one we have for calculating partial orders, that avoids generating combinations (of size K, K+1, K+2,.... N) that contain elements whose coordinates in the sieve matrix is 1.

Reply:
So effectively we push the `for` loop into the `chooseK()`?

Reply:
Yeah, I'm not sure how you would implement such a function without effectively replicating the logic in this method, if I'm understanding the suggestion

Reply:
> So effectively we push the for loop into the chooseK()?

In a sense, yes.

> Yeah, I'm not sure how you would implement such a function without effectively replicating the logic in this method, if I'm understanding the suggestion

The logic will mostly stay the same, but I think we can leverage the pruning of K-sets to prune K+1 in a more efficient, dynamic programming type of, way, instead of recalculating it over and over again as we do here.

We can look at this later though.